bfdriver/src/platform/efi/efi_base/base.c

```diff
@@ -0,0 +1,115 @@
+/*
+ * Bareflank Hypervisor
+ * Copyright (C) 2018 Assured Information Security, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
+ * of the Software, and to permit persons to whom the Software is furnished to do
+ * so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+#include "base.h"
+#include "boot.h"
+#include "mp_service.h"
+#include "bfelf_loader.h"
+#include "common.h"
+
+
+EFI_MP_SERVICES_PROTOCOL *g_mp_services;
+extern char target_vmm_start[];
+extern char target_vmm_end[];
+
+
+VOID __attribute__((ms_abi)) bf_start_hypervisor_on_core(VOID *data)
+{
+    (void)data;
+    _set_ne();
+    Print(L"Starting hypervisor on core.\n");
+    platform_start_core();
+
+    return;
+}
+
+boot_ret_t base_start_fn()
+{
+    EFI_STATUS status;
+
+    uint64_t target_vmm_size = (uint64_t)(target_vmm_end - target_vmm_start);
+    Print(L"target_vmm_size %x\n", target_vmm_size);
+
+    int64_t ret = common_add_module((const char *)target_vmm_start,
+                                    (uint64_t)target_vmm_size);
+    if (ret < 0) {
+        Print(L"common_add_module returned %a\n", ec_to_str(ret));
+        goto fail;
+    }
+
+    Print(L"Loading modules..\n");
+    ret = common_load_vmm();
+    if (ret < 0) {
+        Print(L"common_load_vmm returned %a\n", ec_to_str(ret));
+        goto fail;
+    }
+
+    uint64_t cpus = platform_num_cpus();
+    if (cpus == 0) {
+        Print(L"Error: bf_start_by_startupallaps: bf_num_cpus\n");
+        goto fail;
+    }
+    Print(L"Detected %d CPUs.\n", cpus);
```

Comment:
Same thing.... I feel like there are too many printf statements here which is really noisy.